tfjs-layers/src/layers/nlp/tokenizers_utils.ts

```diff
@@ -141,3 +141,162 @@ export class BytePairTokenizerCache {
+  strs: string[]|string[][],
+  delimRegexPattern: RegExp | string,
+  keepDelimRegexPattern?: RegExp | string): string[][] {
+
+  if (strs[0] instanceof Array) {
+    const mapped = strs.map(arr => regexSplit(
+      arr as string[], delimRegexPattern, keepDelimRegexPattern));
+    return mapped.map((doubleArr) => flatten(doubleArr));
+  }
+
+  strs = strs as string[];
+
+  if (!(delimRegexPattern instanceof RegExp)) {
+    if (keepDelimRegexPattern) {
+      delimRegexPattern = new RegExp(
+        `(${delimRegexPattern})`, 'g');
+    }
+    return strs.map(str => str.split(delimRegexPattern).filter(s => s));
+  }
+
+  let regexPattern = delimRegexPattern as unknown as RegExp;
+  if (!regexPattern.flags.includes('g')) {
+    regexPattern = new RegExp(regexPattern.source, regexPattern.flags + 'g');
+  }
+
+  return strs.map(str => {
+    const matches = str.matchAll(regexPattern);
+
+    const splitString = [];
+    let currIdx = 0;
+    for (const match of matches) {
+      splitString.push(str.slice(currIdx, match.index));
+      if (keepDelimRegexPattern) {
+        splitString.push(
+          str.slice(match.index, match.index! + match[0].length));
+      }
+      currIdx = match.index! + match[0].length;
+    }
+    if (currIdx !== str.length) {
+      splitString.push(str.slice(currIdx, str.length));
+    }
+    return splitString.filter(s => s);
+  });
+}
+
+export async function tensorToStringArr(input: Tensor): Promise<string[]> {
+  return await input.data() as unknown as string[];
+}
+
+export async function tensorArrToString2DArr(
+  inputs: Tensor[]): Promise<string[][]> {
+  return Promise.all(inputs.map(
+    async input => await input.data() as unknown as string[]));
+}
+
+export async function splitStringsForBpe(
+  inputs: Tensor, unsplittableTokens?: string[]): Promise<Tensor[]> {
+
+  // We need to recreate the exact behavior of token presplitting in the
+  // original gpt2 implementation which uses a lookahead. We are using an
+  // alternative by inserting a special token "६" before leading space of
+  // non-space characters and after the trailing space, e.g.,
+  // " tf" will be "६ tf".
+  const pattern1 = new RegExp(`( )([^\s${SPECIAL_WHITESPACES}])`);
+  const pattern2 = new RegExp(`(\s${SPECIAL_WHITESPACES})\$`);
+
+  const inputsStr = (await inputs.data() as unknown as string[]).map(str =>
+    str.replace(pattern1, `६$1$2`).replace(pattern2, `$1६`)
+  );
+
+  let alts: string[];
+  let rawTokens: string[][];
+
+  if (unsplittableTokens && unsplittableTokens.length > 0) {
+    alts = createAltsForUnsplittableTokens(unsplittableTokens);
+    unsplittableTokens.forEach((token, idx) => {
+      const alt = alts[idx];
+      const escapedToken = token.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
+
+      rawTokens = regexSplit(rawTokens !== undefined ?
+        rawTokens : inputsStr, escapedToken, escapedToken);
+      rawTokens = rawTokens.map(
+        arr => arr.map(t => t.replace(escapedToken, alt)));
+    });
```

Comment:
Small Nit: The style guide (go/tsstyle) prefers `for` loops over `forEach`.
```suggestion
    for (const [idx, token] of unsplittableTokens.entries()) {
      const alt = alts[idx];
      const escapedToken = token.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');

      rawTokens = regexSplit(rawTokens !== undefined ?
        rawTokens : inputsStr, escapedToken, escapedToken);
      rawTokens = rawTokens.map(
        arr => arr.map(t => t.replace(escapedToken, alt)));
    }
```